packages/typespec-client-generator-core/src/package.ts

```diff
@@ -247,6 +248,19 @@ function getSdkBasicServiceMethod<TServiceOperation extends SdkServiceOperation>
   const serviceOperation = diagnostics.pipe(
     getSdkServiceOperation<TServiceOperation>(context, operation, methodParameters)
   );
+  // set the correct encode for body parameter according to the content-type
```

Comment:
Another way to set the `encode` is trying to get the http parameter type encode value. That's quite ugly because you have to do a lot of type detection since here we only have `SdkType`.